docs/migration/v12.md

```diff
@@ -0,0 +1,329 @@
+yarn d2-app-scripts migrate js-to-jsx
+```
+
+Check out the [script docs](../scripts/migrate/js-to-jsx.md) for options, tips, and caveats.
+
+## React 18
+
+In terms of the React 18-related changes, upgrading to `@dhis2/cli-app-scripts` version 12 should generally a smooth process for the consuming apps. Most of the [changes](https://react.dev/blog/2022/03/08/react-18-upgrade-guide) shouldn't affect most apps.
+
+### Default props
+
+The main change in React 18 that is expected to affect apps at runtime is that using `MyComponent.defaultProps = { ... }` is deprecated, and will be removed in the next major version. While non-breaking, this change comes along with warning messages in the console for every rendered component that uses them, which can be chatty and annoying.
+
+The fix is to move prop defaults -- either to default parameters for functional components, or to a static property for class components. This code mod can help with the migration, and was used for the `@dhis2/ui` library: [transform.js](https://gist.github.com/kabaros/45ed16660ac89dc48d22e03836b1e981)
+
+**NB!** Note that default params for functional components are computed on each render, so default values like arrays and objects as defaults may unintentionally retrigger `useEffect` and `useMemo` functions on each render. Instead, use a stable reference like so:
+
+```js
+const arr = []
+const MyComponent = ({ options = arr }) => {
+    /* ... */
+}
+```
+
+### Tests
+
+React 18 has minimal effects on the operation of an app in a browser, but it has a more significant effect on tests, depending on what tools are used and how the tests are written.
+
+#### Snapshots
+
+When the tests are up and running, you can expect some small changes from React 18. You may see changed classnames and different whitespace; these shouldn't cause any functional changes, and they're safe to approve.
+
+#### React Testing Library
+
+Testing-library supports React 18 since v13, so the first step is to upgrade that library and the supporting libraries around it. These major versions bumps come with a few breaking changes that we will document here with some examples to make migration easier.
+
+Examples come from the migration for the Aggregate Data Entry app; you can check out the PR [here](https://github.com/dhis2/aggregate-data-entry-app/pull/396) for more examples in context.
+
+##### `userEvent`
+
+`userEvent` is async now in `@testing-library/user-event`. Check the [release notes](https://github.com/testing-library/user-event/releases/tag/v14.0.0) for an idea of the improvements and new API. In practice, this means mostly that you'd need to add an `await` for `userEvent` interactions like `userEvent.click()`.
+
+Actions using the `fireEvent` API should also be migrated to `userEvent`, e.g. `fireEvent.change(selector(), { target: { value: 'input text' } })` should be changed to `userEvent.type(selector(), 'input text')`.
+
+Here are some examples that illustrate the changes:
+
+```diff
```

Comment:
we'll need to do a little testing to see if this will render properly on the developer portal

Reply:
Ah, it doesn't seem to be working :( I copied this into the blog post PR to test:

![Screenshot 2024-11-04 at 14 50 09](https://github.com/user-attachments/assets/1a39ff1a-a9b9-48ec-ab0a-f8fb6699b869)

I'll look into docusaurus support for something similar

Reply:
Ah, just needed to do this: https://github.com/dhis2/developer-portal/pull/167/commits/45c55d803b610be22fe30768662ad5d53b5cbcca

![Screenshot 2024-11-04 at 16 19 48](https://github.com/user-attachments/assets/ab835884-6e95-4522-9707-bb2c12456a49)